File: inflearn_java/src/DFS_BFS/PizzaDistance.java
package DFS_BFS;

import java.util.ArrayList;
import java.util.Scanner;

/**
 * NxN 크기의 도시지도가 있습니다. 도시지도는 1x1 크기의 격자칸으로 이루어져 있습니다. 각 격자칸에는 0은 빈칸, 1은 집, 2는 피자집으로 표현됩니다.
 * 각 격자칸은 좌표(행번호, 열번호)로 표현됩니다. 행번호는 1부터 N까지이고, 열 번호도 1부터 N까지입니다.
 * ...
 */
class Point4 {
    public int x, y;
    Point4 (int x, int y) {
        this.x = x;
        this.y = y;
    }
}

class PizzaDistance {
    static int n, m, len, answer = Integer.MAX_VALUE;
    static int[] combi;
    static ArrayList<Point4> hs, pz;
    public void DFS (int L, int s) {
        if (L == m) {
            int sum = 0;
            for (Point4 h : hs) {
                int dis = Integer.MAX_VALUE;
                for (int i : combi) {
                    dis = Math.min(dis, Math.abs(h.x - pz.get(i).x) + Math.abs(h.y - pz.get(i).y));
                }
                sum += dis;
            }
            answer = Math.min(answer, sum);
        } else {
            for (int i = s; i < len; i++) {
                combi[L] = i;
                DFS(L + 1, i + 1);
            }
        }
    }

    public static void main(String[] args) {
        PizzaDistance T = new PizzaDistance();
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        m = sc.nextInt();
        pz = new ArrayList<>();
        hs = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                int tmp = sc.nextInt();
                if (tmp == 1) hs.add(new Point4(i, j));
                else if (tmp == 2) pz.add(new Point4(i, j));
            }
        }
        len = pz.size();
        combi = new int[m];
        T.DFS(0, 0);
        System.out.println(answer);
    }
}

File: inflearn_python/src/DFS_BFS/PizzaDistance.py
def DFS(L, s):
    global res
    # 조합을 구해주면 된다.
    if L == m:
        # 종착점
        sum = 0 # 도시의 피자 거리
        for j in range(len(hs)): # 0부터 집 개수 순회
            x1 = hs[j][0]
            y1 = hs[j][1]
            # 각 집의 피자 배달 거리
            dis = 2147000000
            for x in cb:
                x2 = pz[x][0]
                y2 = pz[x][1]
                dis = min(dis, abs(x1 - x2) + abs(y1 - y2))

            sum += dis
        if sum < res:
            res = sum

    else:
        for i in range(s, len(pz)):
            cb[L] = i
            DFS(L + 1, i + 1)

if __name__ == "__main__":
    n, m = map(int, input().split()) # n: 도시의 정보, m: 살아남은 피자집의 개수
    board = [list(map(int, input().split())) for _ in range(n)] # 도시의 지도정보
    hs = [] # 집의 좌표
    pz = [] # 피자집의 좌표
    cb = [0] * m # 조합의 경우 수
    res = 2147000000
    for i in range(n):
        for j in range(n):
            if board[i][j] == 1: # 집을 발견하면
                hs.append((i, j))
            elif board[i][j] == 2: # 피자집을 발견하면
                pz.append((i, j))
    DFS(0, 0)
    print(res)